Posts without text or only with emojis are now always accepted in the language check

annando opened the pull request:
When posts are received via a relay server, we can check if they contain some unwanted languages. We now always accept posts without a body (picture posts) or with only emojis in it, since they don't contain the unwanted languages.

annando committed:
Posts without text or only with emojis are now always accepted in the language check

MrPetovan commented on src/Content/Smilies.php:
```diff
@@ -285,4 +285,17 @@ private static function heartReplaceCallback(array $matches): string
+	 */
+	public static function isEmojiPost(string $body): bool
+	{
+		$conv = html_entity_decode(str_replace([' ', "\n", "\r"], '', $body));
```
I think you should do a more comprehensive `preg_replace('#\s#', '', html_entity_decode($body))` to catch potential non-breaking spaces that can be HTML-encoded. `\s` catches all whitespace characters.

A suite of unit tests would be welcome to catch potential corner cases like this one I thought about.

annando replied:
Done

annando committed:
New function to replace blank characters

MrPetovan commented on src/Content/Smilies.php:
```diff
@@ -285,4 +285,17 @@ private static function heartReplaceCallback(array $matches): string
+	 */
+	public static function isEmojiPost(string $body): bool
+	{
+		$conv = html_entity_decode(preg_replace('#\s#', '', $body));
```
This will still be stunted by `&nbsp;` entities that should be decoded first before running through `preg_replace`.
```suggestion
		$conv = preg_replace('#\s#', '', html_entity_decode$body));
```

annando replied:
Done

annando committed:
Move "html_entity_decode"

MrPetovan commented on src/Content/Smilies.php:
```diff
@@ -294,7 +294,7 @@ private static function heartReplaceCallback(array $matches): string
 	public static function isEmojiPost(string $body): bool
 	{
-		$conv = html_entity_decode(preg_replace('#\s#', '', $body));
+		$conv = preg_replace('#\s#', '', html_entity_decode($body));
```
Just tested with `&nbsp;` and it needs this modifier to catch the UTF-8 non-breaking space character in the subject:
```suggestion
		$conv = preg_replace('#\s#u', '', html_entity_decode($body));
```

MrPetovan commented:
I started writing unit tests, and I found out that this emoji doesn't validate the method: 👨‍👨‍👧. This is because it is a composite emoji made of 👨👨 and 👧 (each being 4-byte characters) separated by Zero-Width Joiners characters (encoded on only 3 bytes).

annando committed:
Update src/Content/Smilies.php

Co-authored-by: Hypolite Petovan <hypolite@mrpetovan.com>

MrPetovan reviewed (changes requested):
I started writing unit tests, and I found out that this emoji doesn't validate the method: 👨‍👨‍👧. This is because it is a composite emoji made of 👨👨 and 👧 (each being 4-byte characters) separated by Zero-Width Joiners characters (encoded on only 3 bytes). Back to the drawing board!

MrPetovan committed:
Add unit tests for Smilies::isEmojiPost

- Current implementation is failing tests with emojis including the zero-width-joiner character, encoded on 3 bytes only.

MrPetovan committed:
Add FIXME

- Current implementation is failing tests with emojis including the zero-width-joiner character, encoded on 3 bytes only.

annando commented:
How do we continue with this PR?

MrPetovan commented:
We have to find another way of detecting emoji-only posts, the 4-byte rule isn't universal and will lead to inconsistent results.

annando commented:
Yes. But this one is better than nothing. This PR here just moves the existing check into a separate function. And at the places where the function is used, some false negatives don't really hurt.

MrPetovan commented:
Then we have to make it clear in the method description that it is an approximation, possibly including in the method name, and comment out the failing test cases in order for tests to pass anyway,

annando commented:
We should just put it into the description. Eventually someone will have a look at the function and will improve it.

MrPetovan commented:
"Eventually" is load-bearing here, but the unit test suite I added to your branch shouldn't be failing while we wait for a savior, so it should be commented to fit the present decision.

tobiasd commented:
In case this PR shall be part of the 2023.09 (10) release, please redirect it towards the 2023.09-rc branch.

annando committed:
Merge remote-tracking branch 'upstream/2023.09-rc' into smilies

annando committed:
Use "IntlChar" for the emoji detection

annando commented:
I'm now using the functionality from the `IntlChar` class.

MrPetovan commented:
How do the unit tests look?

annando commented:
I haven't done anything with the tests, so we have to see.

MrPetovan commented:
These can be ran directly with PHPUnit without the whole Continuous Integration infrastructure as they don’t require a database connection, I’ll check out your branch and run the tests manually tonight or tomorrow EST.

MrPetovan commented:
Here are the only two tests cases that are failing:
```
'emojis+HTML else' => [
	'expected' => false,
	'body' => '👀&lt;🤷',
],
'HTML entity else' => [
	'expected' => false,
	'body' => '&deg;',
],
```
`Smilies::isEmojiPost` unexpectedly returns `true` for both cases.

annando commented:
I currently have slight issues with `°`. This is a two byte graphical character that isn't alphanumeric. One could possibly argue, that it can be used to express emotions as well?

MrPetovan commented:
Sure, but in this case you have to be more precise about the meaning of `Smilies::isEmojiPost` up to and including renaming it to match the updated meaning.

I wrote these tests based on its current description, the tests can be changed, but it needs to be accompanied by a change in the method itself, either the code and/or the name and/or the documentation.

annando committed:
Improved Emoji detection

annando commented:
I hope that the added check will now work well for all characters.

MrPetovan commented:
It should, but in the shower I thought about non-Latin scripts like Arabic script, I'm pretty sure it doesn't pass the check.

annando commented:
The function `IntlChar::isalnum` returns "true" for characters in all scripts of the world. I'm using this in the language detection part as well.
